src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSASignatureDeformatter.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography
+{
+    public class DSASignatureDeformatter : AsymmetricSignatureDeformatter
+    {
+        DSA _dsaKey;
+        string _algName;
+
+        public DSASignatureDeformatter()
+        {
+            // The hash algorithm default is SHA1
+            _algName = HashAlgorithmNames.SHA1;
+        }
+
+        public DSASignatureDeformatter(AsymmetricAlgorithm key) : this()
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            _dsaKey = (DSA)key;
+        }
+
+        public override void SetKey(AsymmetricAlgorithm key)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            _dsaKey = (DSA)key;
+        }
+
+        public override void SetHashAlgorithm(string strName)
+        {
+            // The implemenation is symmetric with RSAPKCS1SignatureDeformatter even though _algName
+            // is not used in VerifySignature.
+            try
+            {
+                // Verify the name
+                Oid.FromFriendlyName(strName, OidGroup.HashAlgorithm);
+
+                // Keep the raw strName as Oid may change the case ("SHA1" to "sha1") making it incompatible.
+                _algName = strName;
```

Comment:
http://referencesource.microsoft.com/#mscorlib/system/security/cryptography/dsasignatureformatter.cs,3b245c7de6a74ce8

It has saved to an instance field (which could have been static) the OID for SHA-1.  If CryptoConfig doesn't return that OID for the new input, it throws.

If not for the HashAlgorithm.ToString() wonkiness I'd say the best short term answer is probably something like

```C#
switch (strName?.ToUpperInvariant())
{
    // Definitely these
    case "SHA1":
    case "1.3.14.3.2.26":
    // Maybe these
    case "SHA":
        break;
    default:
        throw new CryptographicUnexpectedException(SR.WhateverTheRightThingIs);
}
```